File: 2015/7/php/main.php
<?php

enum InstructionType {
    case AND;
    case OR;
    case NOT;
    case LSHIFT;
    case RSHIFT;
    case ASSIGN;
}

class Instruction {
    public $operation;
    public $input1;
    public $input2;
    public $output;

    public function __construct($operation, $output, $input1, $input2 = "") {
        $this->operation = $operation;
        $this->input1 = $input1;
        $this->input2 = $input2;
        $this->output = $output;
    }

    public function __toString() {
        return "Instruction: $this->input1 $this->input2 -> $this->output";
    }
}

$instructions = array();
$hashmap = array();

function evaluate($target) {
    global $hashmap;
    global $instructions;
    if (array_key_exists($target, $hashmap)) {
        return $hashmap[$target];
    }

    if (is_numeric($target)) {
        return intval($target);
    }

    $instruction = NULL;
    foreach ($instructions as $i) {
        if ($i->output == $target) {
            $instruction = $i;
            break;
        }
    }

    if ($instruction == NULL) {
        return $target;
    }

    $op1 = is_numeric($instruction->input1) ? intval($instruction->input1) : evaluate($instruction->input1, $hashmap, $instructions);
    $op2 = 0;
    if ($instruction->input2 != "") {
        $op2 = is_numeric($instruction->input2) ? intval($instruction->input2) : evaluate($instruction->input2, $hashmap, $instructions);
    }

    $result = 0;
    switch ($instruction->operation) {
        case InstructionType::AND:
            $result = $op1 & $op2;
            break;
        case InstructionType::OR:
            $result = $op1 | $op2;
            break;
        case InstructionType::NOT:
            $result = ~$op1;
            break;
        case InstructionType::LSHIFT:
            $result = $op1 << $op2;
            break;
        case InstructionType::RSHIFT:
            $result = $op1 >> $op2;
            break;
        case InstructionType::ASSIGN:
            $result = $op1;
            break;
    }

    $hashmap[$target] = $result;
    return $result;
}

function main(){
    global $instructions;
    // get command line arguments
    $args = $_SERVER['argv'];
    if (count($args) < 2) {
        echo "Usage: php main.php <input.txt>\n";
        exit(1);
    }

    $filename = $args[1];

    if (!file_exists($filename)) {
        echo "File not found: $filename\n";
        exit(1);
    }

    $file = fopen($filename, "r");

    if ($file === false) {
        echo "Error opening file: $filename\n";
        exit(1);
    }

    $content = fread($file, filesize($filename));
    fclose($file);

    $lines = explode("\n", $content);

    
    
    foreach ($lines as $line) {
        $words = explode(" ", $line);
        
        if (count($words) == 3){
            $instructions[] = new Instruction(InstructionType::ASSIGN, $words[2], $words[0], "");
        }

        if (count($words) == 4){
            $instructions[] = new Instruction(InstructionType::NOT, $words[3], $words[1], "");
        }

        if (count($words) == 5){
            if ($words[1] == "AND"){
                $instructions[] = new Instruction(InstructionType::AND, $words[4], $words[0], $words[2]);
            }
            if ($words[1] == "OR"){
                $instructions[] = new Instruction(InstructionType::OR, $words[4], $words[0], $words[2]);
            }
            if ($words[1] == "LSHIFT"){
                $instructions[] = new Instruction(InstructionType::LSHIFT, $words[4], $words[0], $words[2]);
            }
            if ($words[1] == "RSHIFT"){
                $instructions[] = new Instruction(InstructionType::RSHIFT, $words[4], $words[0], $words[2]);
            }
        }
    }

    $a = evaluate("a");
    echo "Part 1: $a\n";
    global $hashmap;    
    $hashmap = array("b" => $a);
    $result = evaluate("a");
    echo "Part 2: $result\n";
}

main();
?>

File: 2015/ruby/7/main.rb

$instructions = []
$wires = {}

module InstructionType
    AND = "AND"
    OR = "OR"
    NOT = "NOT"
    LSHIFT = "LSHIFT"
    RSHIFT = "RSHIFT"
    ASSIGN = "ASSIGN"
end

class Instruction
    operation = ""
    input1 = ""
    input2 = ""
    output = ""

    public def initialize(operation, output, input1, input2 = "")
        @operation = operation
        @input1 = input1
        @input2 = input2
        @output = output
    end

    public def operation
        return @operation
    end

    public def input1
        return @input1
    end

    public def input2
        return @input2
    end

    public def output
        return @output
    end
end

def eval(target)
    if $wires[target] != nil
        return $wires[target]
    end

    instruction = $instructions.find { |instruction| instruction.output == target }
    if instruction == nil
        return nil
    end

    op1 = instruction.input1.to_i.to_s == instruction.input1 ? instruction.input1.to_i : eval(instruction.input1)
    op2 = 0
    if instruction.input2 != ""
        op2 = instruction.input2.to_i.to_s == instruction.input2 ? instruction.input2.to_i : eval(instruction.input2)
    end

    if instruction.operation == InstructionType::ASSIGN
        $wires[target] = op1
    end
    if instruction.operation == InstructionType::AND
        $wires[target] = op1 & op2
    end
    if instruction.operation == InstructionType::OR
        $wires[target] = op1 | op2
    end
    if instruction.operation == InstructionType::NOT
        $wires[target] = ~op1
    end
    if instruction.operation == InstructionType::LSHIFT
        $wires[target] = op1 << op2
    end
    if instruction.operation == InstructionType::RSHIFT
        $wires[target] = op1 >> op2
    end
    return $wires[target]
end

def main()    
    # get command line arguments
    args = ARGV
    if args.length < 1
        puts "Usage: ruby main.rb <input.txt>"
        return
    end

    # read file
    file = File.open(args[0], "r")
    if !file
        puts "Failed to open file: #{args[0]}"
        return
    end

    # read file line by line
    file.each_line do |line|
        words = line.split(" ")
        if words.length == 3
            $instructions.push(Instruction.new(InstructionType::ASSIGN, words[2], words[0]))
        end
        if words.length == 4
            $instructions.push(Instruction.new(InstructionType::NOT, words[3], words[1]))
        end
        if words.length == 5
            if words[1] == "AND"
                $instructions.push(Instruction.new(InstructionType::AND, words[4], words[0], words[2]))
            end
            if words[1] == "OR"
                $instructions.push(Instruction.new(InstructionType::OR, words[4], words[0], words[2]))
            end
            if words[1] == "LSHIFT"
                $instructions.push(Instruction.new(InstructionType::LSHIFT, words[4], words[0], words[2]))
            end
            if words[1] == "RSHIFT"
                $instructions.push(Instruction.new(InstructionType::RSHIFT, words[4], words[0], words[2]))
            end
        end
    end

    a = eval("a")
    puts "Part 1: #{a}"
    $wires = {"b" => a}
    a = eval("a")
    puts "Part 2: #{a}"


end

main()